staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go

```diff
@@ -675,16 +693,41 @@ func (c *Cacher) dispatchEvent(event *watchCacheEvent) {
 		// Iterate over watchers interested in exact values for all values.
 		for _, watchers := range c.watchers.valueWatchers {
 			for _, watcher := range watchers {
-				watcher.add(event, c.timer, c.dispatchTimeoutBudget)
+				c.watchersBuffer = append(c.watchersBuffer, watcher)
 			}
 		}
 	}
+	c.Unlock()
+
+	// Since add() can block, we explicitly when cacher is unlocked.
+	for _, watcher := range c.watchersBuffer {
+		watcher.add(event, c.timer, c.dispatchTimeoutBudget)
+	}
+
+	// Finish dispatching phase and stop all watchers that were supposed
```

Comment:
extract all the below into a `stopDispatching` function? Would only make sense if we could have a corresponding `startDispatching` function which was less obvious to me how to separate out.